test/replace-state-patch.html

```diff
@@ -0,0 +1,81 @@
+<script>
+  function __getRunnerWindow() {
+    var lastWindow;
+    var newWindow = window;
+
+    while (lastWindow != newWindow) {
+      lastWindow = newWindow;
+      newWindow = newWindow.parent;
+    }
+
+    return newWindow;
+  };
+
+  var __runnerWindow = __getRunnerWindow();
+  /**
+  * Patches window.History.prorotype.replaceState to count the number of times
+  * it has been called in `__runnerWindow.replaceStateCount`
+  */
+  function patchReplaceState() {
+    if (typeof __runnerWindow.replaceStateCount !== 'number') {
+      __runnerWindow.replaceStateCount = 0;
+    }
+
+    var native = window.History.prototype.replaceState;
+
+    function patch() {
+
+      __runnerWindow.replaceStateCount += 1;
+      return native.apply(window.history, arguments);
+    }
+
+    window.History.prototype.replaceState = patch;
+  }
+
+  var cooldownFunction = function (done) {
+    if (done) { done(); }
+    return Promise.resolve();
+  }
+
+  var isChrome = /Chrome/.test(navigator.userAgent) && /Google Inc/.test(navigator.vendor);
+  var isSafari = /^Apple/.test(navigator.vendor);
+
+  if (isChrome || isSafari) {
+    cooldownFunction = function (done, force) {
+      var resolve;
+      // this is for setups that return promises
+      var cooledDown = new Promise(function (res) { resolve = res });
```

Comment:
Here you're doing both Promise-based logic and callback based logic. Why not write in Promise style, and then have a check at the end like:

```javascript
if (done) {
  cooledDown.then(function() {done();}, function (err) {done(err);});
}
```

Reply:
👍 👍 👍